### ISSUE TITLE
error TS5055: Cannot write file in composite project with specific directory name

### ISSUE THREAD

**Issue by bergundy** [microsoft/TypeScript#46007]
# Bug Report

Hi,

Our composite project (Temporal ~~Node.js~~ TypeScript SDK) build emits TS5055 errors in watch mode after modifying a source file in `packages/lib/worker/` (e.g. `packages/lib/worker/core.ts`).

I've looked at several issues, [the docs](https://www.typescriptlang.org/docs/handbook/project-references.html) and google searches and could not find an explanation for this.

I apologize in advance for not finding a minimal repro.

### 🔎 Search Terms

- TS5055
- overwrite input file
- cannot write file

### 🕗 Version & Regression Information

`4.4.3` but happens with any `4.x.x` version I've tried (haven't tried earlier versions).

### 💻 Code

I tried to create a simple repro but I couldn't, unfortunately this project is a bit complex to setup (see https://github.com/temporalio/sdk-node/blob/main/CONTRIBUTING.md)
Hopefully the hint below will be enough to skip the repro.

### 🙁 Actual behavior

To reproduce, clone the repo from this commit:
https://github.com/temporalio/sdk-node/tree/9592be12e66af2939d5f5d71984007d8a95a2a76

- Set up the project as instructed above.
- Run `npm run build.watch` in a separate terminal
- Change `packages/worker/lib/core.ts`
- Observe the emitted errors:

```
@temporalio/test: error TS5055: Cannot write file '/Users/bergundy/temporal/sdk-node/packages/worker/lib/core.d.ts' because it would overwrite input file.
@temporalio/test: error TS5055: Cannot write file '/Users/bergundy/temporal/sdk-node/packages/worker/lib/dependencies.d.ts' because it would overwrite input file.
@temporalio/test: error TS5055: Cannot write file '/Users/bergundy/temporal/sdk-node/packages/worker/lib/errors.d.ts' because it would overwrite input file.
@temporalio/test: error TS5055: Cannot write file '/Users/bergundy/temporal/sdk-node/packages/worker/lib/index.d.ts' because it would overwrite input file.
@temporalio/test: error TS5055: Cannot write file '/Users/bergundy/temporal/sdk-node/packages/worker/lib/interceptors.d.ts' because it would overwrite input file.
@temporalio/test: error TS5055: Cannot write file '/Users/bergundy/temporal/sdk-node/packages/worker/lib/isolate-builder.d.ts' because it would overwrite input file.
@temporalio/test: error TS5055: Cannot write file '/Users/bergundy/temporal/sdk-node/packages/worker/lib/isolate-context-provider.d.ts' because it would overwrite input file.
@temporalio/test: error TS5055: Cannot write file '/Users/bergundy/temporal/sdk-node/packages/worker/lib/logger.d.ts' because it would overwrite input file.
@temporalio/test: error TS5055: Cannot write file '/Users/bergundy/temporal/sdk-node/packages/worker/lib/server-options.d.ts' because it would overwrite input file.
@temporalio/test: error TS5055: Cannot write file '/Users/bergundy/temporal/sdk-node/packages/worker/lib/worker.d.ts' because it would overwrite input file.
@temporalio/test: error TS5055: Cannot write file '/Users/bergundy/temporal/sdk-node/packages/worker/lib/workflow.d.ts' because it would overwrite input file.
@temporalio/test: 5:44:49 PM - Found 11 errors. Watching for file changes.
```

Strangely, if I rename the `worker` directory to `workerr` **this does not happen**.
This commit includes the rename: https://github.com/temporalio/sdk-node/commit/671b9c2ad0ae16d87e198d38ed5aba54e631848e

### 🙂 Expected behavior

No build errors should be emitted.

**Comment by MartinJohns** [comment]
What makes you think this is an issue with TypeScript? Have you tried this without the massive amount of tooling on top?

**Comment by andrewbranch** [comment]
Does this happen if you just run `tsc --build --watch` on the `test` package without simultaneously running the other `build.watch` scripts?

**Comment by bergundy** [comment]
@andrewbranch Yes, this happens if I only run `(cd packages/test && npx tsc --build --watch)`
@MartinJohns are you referring to npm scripts or lerna? Those are the only tools we have on top both pretty standard.

**Comment by MyGitHup1** [comment]
If you turn on declaration `declaration:true` please try turning off declaration `declaration: false` might be ok

**Comment by bergundy** [comment]
@MyGitHup1 I can't turn declaration off because it's a composite project.
```
../worker/tsconfig.json(5,5): error TS6304: Composite projects may not disable declaration emit.
```

**Comment by bergundy** [comment]
I found the cause of this issue.

We somehow accidentally added a circular import in a `.d.ts` file referencing a type from the `@temporalio/worker` package.

What's strange is that there was no error emitted by TypeScript, I think we may need to make `index.d.ts` checked by the compiler and added to the composite project setup.
I'm not sure how this should be done though since there's nothing to compile in that package, there's only a single `.d.ts` file.

I've created a minimal reproduction of my issue here: https://github.com/bergundy/ts5055-repro
I'd say this might be a TypeScript issue because there's no assistance from the compiler indicating that something is wrong.

```shellsession
$ npm run build.watch

[5:02:19 AM] Starting compilation in watch mode...

error TS5055: Cannot write file '.../TS5055-repro/packages/b/lib/index.d.ts' because it would overwrite input file.

[5:02:19 AM] Found 1 error. Watching for file changes.
```

**Comment by sandersn** [comment]
@bergundy should this issue still be open then? Or do we need to have a different issue tracking the lack of error from Typescript?

**Comment by bergundy** [comment]
I've figured out what I did wrong but I think there might be an issue with the TypeScript not helping me figure this out.
I've managed to find out the reason because a user of my library used it with `skipLibCheck: false` and I think also `node_modules` was not excluded.

I've been using https://www.npmjs.com/package/@tsconfig/node14 as my base tsconfig which sets `skipLibCheck` to `true` which seems to be recommended for performance reasons, setting it to `false` indeed fixes the experience and `tsc` emits a meaningful message.
